### PR TITLE
Fix UB in string comparisons, IO, and hashing

### PR DESCRIPTION
**Pull request by jabraham17**
Fixes several issues in our String and IO code where certain edge cases could result in C undefined behavior, leading to miscompilation with some backend compilers

* fixes string comparisons on empty strings, due to an issue where empty strings are not fully allocated
    * this resulted in null pointers being passed to `memcmp`, which is UB
* fixes an issue where a IO read of length 0 would result in a null pointer being passed to `memcpy`, which is UB
* fixes signed int overflow in a hash function, caused by using signed ints instead of unsigned ints

I also added various bounds checks and asserts to get real errors for these silent UB errors in the future

Resolves https://github.com/chapel-lang/chapel/issues/28224

- [x] paratest with/without gasnet
- [x] check that reproducer from https://github.com/chapel-lang/chapel/issues/28224 now works

[Reviewed by @DanilaFe]

### CODE REVIEW

File: modules/internal/ByteBufferHelpers.chpl [hunk 1]
```diff
@@ -164,14 +164,20 @@ module ByteBufferHelpers {
 
   private inline proc _strcmp_local(buf1: bufferType, len1: int,
                                     buf2: bufferType, len2: int) : int {
+    // these 3 conditionals prevent edge cases when a string buffer might be unallocated
+    // this can occur with empty strings like `""`
+    // see test/types/string/emptyString.chpl for an example
+    if (len1 == 0) & (len2 == 0) then return 0;
```
Comment (DanilaFe) [inline modules/internal/ByteBufferHelpers.chpl:170]: Why use non-short-circuiting `&` here?

File: modules/standard/IO.chpl [hunk 2]
```diff
@@ -7165,7 +7163,12 @@ proc readStringBytesData(ref s: ?t /*: string or bytes*/,
   // in the event that s.type == string.
 
   var len:c_ssize_t = nBytes.safeCast(c_ssize_t);
-  var err = qio_channel_read_amt(false, _channel_internal, sLocal.buff, len);
+  var err;
+  if len != 0 {
+    err = qio_channel_read_amt(false, _channel_internal, sLocal.buff, len);
+  } else {
+    err = 0:errorCode;
+  }
```
Comment (DanilaFe) [inline modules/standard/IO.chpl:7171]: why not `var err: errorCode` and later `err = 0`?

File: modules/standard/OS.chpl [hunk 3]
```diff
@@ -1092,6 +1099,13 @@ module OS {
     */
     pragma "fn synchronization free"
     inline proc memcpy(dest:c_ptr(void), const src:c_ptr(void), n: c_size_t) {
+      if chpl_checkNilDereferences {
+        import HaltWrappers;
+        if dest == nil then
+          HaltWrappers.nilCheckHalt("dest argument to memcpy is nil");
+        if src == nil then
+          HaltWrappers.nilCheckHalt("src argument to memcpy is nil");
```
Comment (DanilaFe) [inline modules/standard/OS.chpl]: This could be a helper function, inline if that worries you.

**Review by DanilaFe — APPROVED** [review]
I'd be curious to see if there are performance implications here. We are introducing a decent amount of new branching.

File: modules/internal/ByteBufferHelpers.chpl (hunk 1 above)
Comment (jabraham17) [reply modules/internal/ByteBufferHelpers.chpl:170]: Since this is likely performance sensitive code, I want to avoid the branching overhead of `&&`

File: modules/standard/IO.chpl (hunk 2 above)
Comment (jabraham17) [reply modules/standard/IO.chpl:7171]: No strong reason, just a different style. I can switch it

File: modules/standard/OS.chpl (hunk 3 above)
Comment (jabraham17) [reply modules/standard/OS.chpl]: If I was to introduce a helper, it would be 
```
if chpl_checkNilDereferences then {
  doUBChecking(dest, src)
}
```
and not inline the function

Comment (DanilaFe) [reply modules/standard/OS.chpl]: you could define `doUBChecking` with a `where` clause on `chpl_checkNilDereferences`. When marked inline, the resolved overload (empty when bounds checking is off) will just disappear, no?

I'd probably call it "checkSourceAndDestBuffer", but that's a nit.

Comment (jabraham17) [reply modules/standard/OS.chpl]: The where clause isn't needed, the reason I prefer `if then notInlinedCheckingFunc` is that with `--no-checks`, `chpl_checkNilDereferences` will get param folded out and then `notInlinedCheckingFunc` is not resolved. With `--checks`, `notInlinedCheckingFunc` being inlined will increase code size unnecessarily

**Comment by bradcray** [comment]
Nice, thanks!

**Comment by benharsh** [comment]
Already paying off! The ``chpl-mem.h`` assertion just caught a bug with multiplying a string by ``0`` and trying to write the result.
